### PR TITLE
Record shape and type in autograd to validate gradients

### PR DESCRIPTION
**Pull request by colesbury**
In the autograd engine, we now validate:

 - The gradient's shape matches the expected shape
 - The gradient's type is compatible with the expected type (i.e. exactly the same or a sparse version of the expected type)

We can also check that gradients are defined when we expect them to be. This check is currently commented out because it triggers in TestJit.test_ge_optimized.

### CODE REVIEW

**Review by goldsborough — APPROVED** [review]
Approve for C++ API parts

File: torch/csrc/autograd/function.h [hunk 1]
```diff
@@ -123,20 +121,37 @@ struct Function : std::enable_shared_from_this<Function> {
...
+
+  // Marker for expected undefined input
+  struct undefined_input {};
 
   /// Increments the number of inputs of the function and returns the previous
-  /// value.
-  uint32_t bump_inputs() noexcept {
-    return num_inputs_++;
+  /// value. Type and shape are the expected properties of the input.
+  uint32_t bump_inputs(const at::Type& type, at::IntList shape) noexcept {
```
Comment (ezyang) [inline torch/csrc/autograd/function.h]: A bit of a misnomer now, no?

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -288,6 +288,47 @@ static variable_list call_post_hooks(Function& fn, variable_list outputs, variab
   return outputs;
 }
 
+static bool is_compatible_type(const at::Type& expected, const at::Type& actual) {
```
Comment (ezyang) [inline torch/csrc/autograd/engine.cpp]: Worth a comment saying that it's OK for a dense tensor to have a sparse gradient, but not vice versa.

**Review by ezyang — APPROVED** [review]
Quality work, though I'm surprised at all the places you needed to add `bump_inputs`, where there wasn't previously a `set_num_inputs`. Intruiging indeed.

File: tools/autograd/templates/VariableType.cpp [hunk 3]
```diff
@@ -359,35 +359,35 @@ static void throw_error_out_requires_grad(const char* name) {
 
 static void rebase_history(Variable& var, std::shared_ptr<Function> grad_fn) {
   if (grad_fn && var.defined()) {
-    grad_fn->set_num_inputs(1);
+    grad_fn->bump_inputs(var.type(), var.sizes());
```
[the remark this answers is not in the document]
Comment (colesbury) [reply tools/autograd/templates/VariableType.cpp]: The `grad_fn` is always a new object or null in `rebase_history` and `set_history`. (The `var` may not be new in the case of in-place ops).

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 4]
```diff
@@ -13,12 +13,13 @@ using at::Tensor;
...
-// AccumulateGrad sets sequence_nr to the max value so it's always called 
+// AccumulateGrad sets sequence_nr to the max value so it's always called
 // ASAP during backwards.
 AccumulateGrad::AccumulateGrad(Variable variable_)
-    : Function(/*num_inputs=*/1
-              , /*sequence_nr=*/UINT64_MAX)
-              , variable(std::move(variable_)) {}
+    : Function(/*sequence_nr=*/UINT64_MAX)
+    , variable(std::move(variable_)) {
+  add_input_metadata(variable.type(), variable.sizes());
```
Comment (apaszke) [inline torch/csrc/autograd/functions/accumulate_grad.cpp]: Is this correct? `AccumulateGrad` shouldn't return anything, so it doesn't have any inputs

**Review by apaszke — COMMENTED** [review]
FWIW, the lack of checks enabled some workarounds that greatly improve the perf for computing a Jacobian (see e.g. #7786) and I'm a bit sad to see it go 😕

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 4 above)
Comment (colesbury) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: The input to AccumulateGrad is the gradient. It has no outputs. (The previous code had `AccumulateGrad(...)  : Function(/*num_inputs=*/1, ...)`

**Comment by colesbury** [comment]
@apaszke, passing in a larger grad_output only ever worked in trivial cases

**Comment by apaszke** [comment]
@colesbury it didn't work with library ops, but the solution was to implement custom Functions in Python that would account for an extra batch dimension in the gradient.

File: torch/csrc/autograd/functions/accumulate_grad.cpp (hunk 4 above)
Comment (apaszke) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: Ohhh right. `input_metadata` is for forward outputs (i.e. it's written from the perspective of the backward pass). Never mind the comment

**Comment by colesbury** [comment]
@apaszke I guess we could expose `input_metadata` to Python autograd function if we want to support that sort of trick. i.e. if you want implement those sorts of functions you could call something like `ctx.clear_input_metadata()` to avoid the shape and type checks.

**Comment by apaszke** [comment]
@colesbury makes sense. It would be good to have that
